pkg/oc/cli/admin/mustgather/mustgather.go

```diff
@@ -0,0 +1,330 @@
+	cmd.Flags().BoolVar(&o.Keep, "keep", false, "Do not delete temporary resources when command completes.")
+	_ = cmd.Flags().MarkHidden("keep")
+
+	return cmd
+}
+
+func (o *Options) Complete(f kcmdutil.Factory, cmd *cobra.Command, args []string) error {
+	if i := cmd.ArgsLenAtDash(); i != -1 && i < len(args) {
+		o.Command = args[i:]
+	} else {
+		o.Command = args
+	}
+
+	var err error
+
+	if o.Config, err = f.ToRESTConfig(); err != nil {
+		return err
+	}
+
+	if o.Client, err = kubernetes.NewForConfig(o.Config); err != nil {
+		return err
+	}
+
+	if len(o.DestDir) == 0 {
+		o.DestDir = fmt.Sprintf("must-gather.local.%06d", rand.Int63())
+	}
+
+	if len(o.Image) == 0 {
+		imageClient, err := v1.NewForConfig(o.Config)
+		if err != nil {
+			return err
+		}
+		imageStream, err := imageClient.ImageStreams("openshift").Get("must-gather", metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		var ok bool
+		if o.Image, ok = util.ResolveLatestTaggedImage(imageStream, "latest"); !ok {
+			return fmt.Errorf("unable to to resolve must-gather image")
+		}
+	}
+
+	o.RsyncRshCmd = rsync.DefaultRsyncRemoteShellToUse(cmd.Parent())
+
+	return nil
+}
+
+type Options struct {
+	genericclioptions.IOStreams
+
+	Config *rest.Config
+	Client kubernetes.Interface
+
+	NodeName string
+	DestDir  string
+	Image    string
+	Command  []string
+	Keep     bool
+
+	RsyncRshCmd string
+}
+
+// Run creates and runs a must-gather pod.d
+func (o *Options) Run(rsyncCmd *cobra.Command) error {
+	if len(o.Config.BearerToken) != 0 {
+		return fmt.Errorf("cannot run must-gather with a token based kubeconfig")
+	}
+	if len(o.Image) == 0 {
+		return fmt.Errorf("missing an image")
+	}
+
+	var err error
+
+	// create namespace
+	ns, err := o.Client.CoreV1().Namespaces().Create(&corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "openshift-must-gather-",
+			Labels: map[string]string{
+				"openshift.io/run-level": "0",
+			},
+			Annotations: map[string]string{
+				"oc.openshift.io/command": "oc adm must-gather",
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if !o.Keep {
+		defer func() {
+			err = o.Client.CoreV1().Namespaces().Delete(ns.Name, nil)
+		}()
+	} else {
+		fmt.Fprintf(o.Out, "namespace/%s created.\n", ns.Name)
+	}
+
+	clusterRoleBinding, err := o.Client.RbacV1().ClusterRoleBindings().Create(o.newClusterRoleBinding(ns.Name))
+	if err != nil {
+		return err
+	}
+	if !o.Keep {
+		defer func() {
+			err = o.Client.RbacV1().ClusterRoleBindings().Delete(clusterRoleBinding.Name, &metav1.DeleteOptions{})
+		}()
+	} else {
+		fmt.Fprintf(o.Out, "clusterrolebinding.rbac.authorization.k8s.io/%s created.\n", clusterRoleBinding.Name)
+	}
+
+	// create pod
+	pod, err := o.Client.CoreV1().Pods(ns.Name).Create(o.newPod(o.NodeName))
+	if err != nil {
+		return err
+	}
+
+	// wait for pod to be running (gather has completed)
+	podPhase, err := o.waitForPodNotPending(pod)
+	if err != nil {
+		return err
+	}
+	if podPhase != corev1.PodRunning {
```

Comment:
Why this is not part of the `waitForPodNotPending` ?

Reply:
This is still not addressed.

Reply:
I guess there was a earlier version when I cared about phase being 'Running' vs 'Succeeded' vs. `Unknown`, etc..  I don't see anything wrong with the way it is now, but I can refactor if there is time.